File: javascript/lru_cache.js
'use strict';

function Node(key, value, prev = null, next = null) {
  this.key = key;
  this.value = value;
  this.prev = prev;
  this.next = next;
}

/**
* @param {number} capacity
*/
var LRUCache = function (capacity) {
  this.capacity = capacity;

  // Hashtable for tracking which items are currently in the cache
  this.keys = {};

  // Pointers for linked list (head = most recently accessed)
  this.head = null;
  this.tail = null;
};

/** 
 * Return the key's value or -1 if it doesn't exist in the cache
* @param {number} key
* @return {number}
*/
LRUCache.prototype.get = function (key) {
  console.log('get', key);
  if (!(key in this.keys)) {
    return -1;
  }

  // Get the node associated with the key in order to get its value
  const { value } = this.keys[key];
  // Remove the node from its current position and promote it to the head of the list
  this.remove(this.keys[key]);
  const newHead = this.push(key, value);
  this.keys[key] = newHead;

  return value;
};

/** 
* @param {number} key 
* @param {number} value
* @return {void}
*/
LRUCache.prototype.put = function (key, value) {
  console.log('put', key, value);
  if (this.capacity < 1) {
    return;
  }

  // If the item is already in the cache, remove the old version of it
  if (key in this.keys) {
    const targetNode = this.keys[key];
    this.remove(targetNode);
  } else if (Object.keys(this.keys).length > this.capacity) {
    // If the item is not in the cache but we're at capacity, remove the LRU item
    const nodeToDelete = this.pop();
    delete this.keys[nodeToDelete.key];
  }

  // Otherwise just add the node to the head and update hashtable
  const newHead = this.push(key, value);
  this.keys[key] = newHead;
};

/**
 * Add a new node to the head of the cache's linked list
 * @param {number} key 
 * @param {number} value 
 * @return {Node}
 */
LRUCache.prototype.push = function (key, value) {
  const newHead = new Node(
    key,
    value,
    null,
    this.head
  );

  if (!this.head) {
    this.tail = newHead;
  } else {
    const oldHead = this.head;
    newHead.next = oldHead;
    oldHead.prev = newHead;
  }

  this.head = newHead;
  return newHead;
}

/**
 * Remove the node at the tail of the cache's linked list and return its 'key' value
 * @return {Node} 
 */
LRUCache.prototype.pop = function () {
  const oldTail = this.tail;

  if (this.head.key === this.tail.key) {
    this.head = null;
    this.tail = null;
  } else {
    this.tail = this.tail.prev;;
    this.tail.next = null;
  }

  return oldTail;
}

/**
 * 
 * @param {Node} node 
 */
LRUCache.prototype.remove = function (node) {
  const { prev, next } = node;
  console.log('remove', node, prev, next);

  if (node.key === this.head.key && node.key === this.tail.key) {
    // If there's only one item in the list, reset head and tail
    this.head = null;
    this.tail = null;
  } else if (node.key === this.head.key) {
    this.head = next;
    this.head.prev = null;
  } else if (node.key === this.tail.key) {
    this.tail = prev;
    this.tail.next = null;
  } else {
    node.prev.next = next;
    node.next.prev = prev;
  }
}

/** 
* Your LRUCache object will be instantiated and called as such:
* var obj = Object.create(LRUCache).createNew(capacity)
* var param_1 = obj.get(key)
* obj.put(key,value)

const cache = new LRUCache(2);

cache.put(1, 1);
cache.put(2, 2);
cache.get(1);       // returns 1
cache.put(3, 3);    // evicts key 2
cache.get(2);       // returns -1 (not found)
cache.put(4, 4);    // evicts key 1
cache.get(1);       // returns -1 (not found)
cache.get(3);       // returns 3
cache.get(4);       // returns 4
*/

File: python/lru_cache.py
# https://leetcode.com/problems/lru-cache/description/

class Node:
    
    def __init__(self, key, value, prev = None, next = None):
        self.key = key
        self.value = value
        self.prev = prev
        self.next = next

class LRUCache:

    def __init__(self, capacity):
        self.capacity = capacity
        self.key_dict = {}
        self.head = None
        self.tail = None
        
    # Get the value (will always be positive) of the key if the key exists in the cache, otherwise return -1.
    def get(self, key):
        if key not in self.key_dict:
            return -1
        value = self.key_dict[key].value
        self.delete_by_key(key)
        new_node = self.add_to_head(key, value)
        self.key_dict[key] = new_node
        return value
           
    # Set or insert the value if the key is not already present.
    # When the cache reached its capacity, it should invalidate the least recently used item before inserting a new item.
    def put(self, key, value):
        if self.capacity < 1:
            return
        # 1. Not at capacity, key isn't already in the list -> add it
        # 2. Not at capacity, key is already in the list -> pop old key off, insert at head
        #    At capacity, key is already in the list -> pop old key off, insert with new value at head
        # 3. At capacity, key isn't already in the list -> pop LRU off, insert at head
        if key in self.key_dict:
            self.delete_by_key(key)
        elif len(self.key_dict) >= self.capacity:
            key_to_delete = self.delete_from_tail()
            del self.key_dict[key_to_delete]
        
        new_node = self.add_to_head(key, value)
        self.key_dict[key] = new_node

    def add_to_head(self, key, value):
        new_node = Node(key, value)
        if self.head == None:
            self.head = new_node
            self.tail = new_node
        else:
            old_head = self.head
            new_node.next = old_head
            old_head.prev = new_node
            self.head = new_node
    
        return new_node 

    def delete_from_tail(self):
        key_to_delete = self.tail.key
        if self.head == self.tail:
            self.head, self.tail = None, None
        else:
            self.tail = self.tail.prev
            self.tail.next = None
           
        return key_to_delete
    
    def delete_by_key(self, key):
        node = self.key_dict[key]
        node_prev = node.prev
        node_next = node.next
        if node == self.head and node == self.tail:
            self.head, self.tail = None, None
        elif node == self.head:
            node_next.prev = None
            self.head = node_next
        elif node == self.tail:
            node_prev.next = None
            self.tail = node_prev
        else:
            node_prev.next = node_next
            node_next.prev = node_prev

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)